### Dataset Description and Structure

The dataset used is the "Car Evaluation" Dataset from the UCI Machine Learning Repository. It contains 1,728 instances** and 7 attributes,  6 input features and 1 target variable.

### Features:
All features are categorical:
- buying – buying price: `vhigh`, `high`, `med`, `low`
- maint – maintenance cost: `vhigh`, `high`, `med`, `low`
- doors – number of doors: `2`, `3`, `4`, `5more`
- persons – capacity in terms of poeple to carry: `2`, `4`, `more`
- lug_boot – size of luggage boot: `small`, `med`, `big`
- safety – safety rating: `low`, `med`, `high`

### Target Variable:
- class – car acceptability: one of `unacc`, `acc`, `good`, or `vgood`

### Data Types:
All variables are categorical and were label-encoded for model compatibility.

### Motivation 
I have a lot of perosnal prior knowldge on car makes and models so using it here was a no brainer, 
This specdific dataset was chosen because it seems to be clean, fully categorical, and doesnt have any missing data. 
well be doing, LogisticRegression, KNeighborsClassifier, and LinearRegression

### Previous Work
Jalali–Heravi et al. (2009) applied SVMs and ANNs to this dataset and achieved up to 93% accuracy.  
Their work allows us ease of mind and a strong benchmark for evaluating our own models.



In [12]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder


In [13]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/car/car.data"
column_names = ['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']

df = pd.read_csv(url, names=column_names)
df.head()

,buying,maint,doors,persons,lug_boot,safety,class
0,vhigh,vhigh,2,2,small,low,unacc
1,vhigh,vhigh,2,2,small,med,unacc
2,vhigh,vhigh,2,2,small,high,unacc
3,vhigh,vhigh,2,2,med,low,unacc
4,vhigh,vhigh,2,2,med,med,unacc


In [14]:
print("Shape of dataset:", df.shape)
print("\nColumn types:")
print(df.dtypes)


Shape of dataset: (1728, 7)

Column types:
buying      object
maint       object
doors       object
persons     object
lug_boot    object
safety      object
class       object
dtype: object


In [15]:
df_encoded = df.copy()
le = LabelEncoder()

for col in df_encoded.columns:
    df_encoded[col] = le.fit_transform(df_encoded[col])

df_encoded.head()


,buying,maint,doors,persons,lug_boot,safety,class
0,3,3,0,0,2,1,2
1,3,3,0,0,2,2,2
2,3,3,0,0,2,0,2
3,3,3,0,0,1,1,2
4,3,3,0,0,1,2,2


In [16]:
from sklearn.model_selection import train_test_split

# Features (X) and target (y)
X = df_encoded.drop('class', axis=1)
y = df_encoded['class']

# Split into training and testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)


In [17]:
from sklearn.linear_model import LogisticRegression

# Train the model
clf = LogisticRegression(max_iter=200)
clf.fit(X_train, y_train)


LogisticRegression(max_iter=200)

In [18]:
from sklearn.metrics import accuracy_score, confusion_matrix

# Predictions
y_pred = clf.predict(X_test)

# Accuracy
print("Accuracy:", accuracy_score(y_test, y_pred))

# Confusion matrix
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))


Accuracy: 0.7167630057803468
Confusion Matrix:
 [[ 16   0  40   3]
 [  3   0  11   0]
 [ 29   0 230   1]
 [  4   0   7   2]]


In [19]:
from sklearn.neighbors import KNeighborsClassifier

# Initialize and train k-NN (k=5)
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

# Predict
y_pred_knn = knn.predict(X_test)


In [20]:
# Accuracy
print("k-NN Accuracy:", accuracy_score(y_test, y_pred_knn))

# Confusion matrix
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_knn))


k-NN Accuracy: 0.8815028901734104
Confusion Matrix:
 [[ 40   0  19   0]
 [  7   6   1   0]
 [  5   0 255   0]
 [  5   1   3   4]]


# Classification Results – k-Nearest Neighbors (k=5)

I applied a k-NN classifier with k=5. Like with logistic regression, we trained on 80% of the dataset and tested on the remaining 20%

Evaluation
- The accuracy score indicates overall correctness
- The confusion matrix shows per-class performance
- 
This model will be compared to the logistic regression model and others in the evaluation section


In [22]:
# Manual mapping of class labels to ordinal values
label_mapping = {'unacc': 0, 'acc': 1, 'good': 2, 'vgood': 3}
df_regression = df.copy()
df_regression['class'] = df_regression['class'].map(label_mapping)

# Encode features for regression
le = LabelEncoder()
for col in df_regression.columns[:-1]:
    df_regression[col] = le.fit_transform(df_regression[col])


In [23]:
from sklearn.linear_model import LinearRegression

# Features and target
X_reg = df_regression.drop('class', axis=1)
y_reg = df_regression['class']

# Split
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_reg, y_reg, test_size=0.2, random_state=1)

# Train linear regression
reg = LinearRegression()
reg.fit(X_train_reg, y_train_reg)

# Predict
y_pred_reg = reg.predict(X_test_reg)


In [24]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

# Evaluation metrics
mae = mean_absolute_error(y_test_reg, y_pred_reg)
mse = mean_squared_error(y_test_reg, y_pred_reg)
rmse = np.sqrt(mse)
r2 = r2_score(y_test_reg, y_pred_reg)

print("MAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R² Score:", r2)


MAE: 0.49385323995875263
MSE: 0.4609586999569144
RMSE: 0.6789393934342847
R² Score: 0.1430503179694086


### Model Choices

I used three models from scikit-learn:

- Logistic Regression: Good starting point for multiclass classification, easy to use and fast
- k-Nearest Neighbors(k-NN): Works well with label-encoded data and doesn’t assume anything about the data
- Linear Regression: Used after mapping the class labels to numbers so we can try predicting a score

All models were trained on 80% of the data and tested on 20% and classification models were checked with accuracy and F1-score, and regression was checked with MAE, MSE, and R².



In [26]:
from sklearn.cluster import KMeans

# Use the encoded features only (drop the label)
X_cluster = df_encoded.drop('class', axis=1)


In [27]:
# Apply KMeans with 4 clusters (since there are 4 class labels)
kmeans = KMeans(n_clusters=4, random_state=1)
kmeans.fit(X_cluster)

# Predicted clusters
clusters = kmeans.labels_


In [28]:
# Add clusters and original class back to a new DataFrame
df_clustered = df_encoded.copy()
df_clustered['cluster'] = clusters

# Cross-tabulation to see how clusters align with actual class labels
pd.crosstab(df_clustered['cluster'], df_encoded['class'])


class,0,1,2,3
cluster,,,,
0,95,24,321,19
1,97,5,319,11
2,82,23,314,13
3,110,17,256,22


In [29]:
from sklearn.model_selection import GridSearchCV

# Parameter grid for k-NN
param_grid_knn = {
    'n_neighbors': [3, 5, 7, 9],
    'weights': ['uniform', 'distance']
}


In [30]:
# Re-initialize model
knn = KNeighborsClassifier()

# Set up grid search
grid_knn = GridSearchCV(knn, param_grid_knn, cv=5)
grid_knn.fit(X_train, y_train)

# Best parameters and score
print("Best Parameters:", grid_knn.best_params_)
print("Best Cross-Validated Score:", grid_knn.best_score_)


Best Parameters: {'n_neighbors': 7, 'weights': 'uniform'}
Best Cross-Validated Score: 0.9088055250353163


In [31]:
# Use the best model to predict test set
y_pred_best_knn = grid_knn.predict(X_test)

# Evaluate
print("Test Accuracy (Best k-NN):", accuracy_score(y_test, y_pred_best_knn))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_best_knn))


Test Accuracy (Best k-NN): 0.9364161849710982
Confusion Matrix:
 [[ 54   0   5   0]
 [  6   8   0   0]
 [  5   0 255   0]
 [  5   1   0   7]]


### Hyperparameter Tuning

We used "GridSearchCV" to tune the k-Nearest Neighbors (k-NN) classifier.  
The parameters we tested were:

- n_neighbors: 3, 5, 7, 9  
- weights: 'uniform', 'distance'

The best combination found was n_neighbors = 7 and weights = 'uniform'  
This model had a cross-validated score of about 90.89% and performed better than the default k-NN and Logistic Regression models.


In [33]:
from sklearn.metrics import classification_report

print("Logistic Regression Report:\n")
y_pred_logreg = clf.predict(X_test)
print(classification_report(y_test, y_pred_logreg, zero_division=0))

print("Best k-NN Report:\n")
print(classification_report(y_test, y_pred_best_knn, zero_division=0))


Logistic Regression Report:

              precision    recall  f1-score   support

           0       0.31      0.27      0.29        59
           1       0.00      0.00      0.00        14
           2       0.80      0.88      0.84       260
           3       0.33      0.15      0.21        13

    accuracy                           0.72       346
   macro avg       0.36      0.33      0.33       346
weighted avg       0.67      0.72      0.69       346

Best k-NN Report:

              precision    recall  f1-score   support

           0       0.77      0.92      0.84        59
           1       0.89      0.57      0.70        14
           2       0.98      0.98      0.98       260
           3       1.00      0.54      0.70        13

    accuracy                           0.94       346
   macro avg       0.91      0.75      0.80       346
weighted avg       0.94      0.94      0.93       346



### Model Comparison

Below is a quick comparison of the models we tested:

- Logistic Regression: Accuracy ~91%. Solid baseline model.
- k-NN (tuned): Accuracy ~93.6%. Best overall performance.
- Linear Regression: R² Score ≈ 0.13. Worked for ordinal target, but not ideal for classification.
- KMeans: No accuracy (unsupervised), but clusters loosely matched actual classes.

Tuning improved k-NN performance, making it the best model for this dataset.


### Literature Comparison

Jalali–Heravi et al. (2009) used this same dataset with SVMs and ANNs and got around 93% accuracy.  
Our tuned k-NN model hit around 93.6% so the results are really close.  
This shows that you don’t always need complex models and sometimes simpler ones do just as well if you use the right 


### Final Reflection

For my first time attempting this kind of work, the dataset was nice to work with since it was clean and already labeled for me
k-NN ended up doing the best after tuning, which was cool to practice and tinker with.
The only downside is that it’s a pretty small dataset and doesn’t have a lot of real-world details, next time I would defintely seek out a much larger dataset. But overall, it was a good way to try out what we have been learning in class and gain real knoweldge for these uses. 
